parser/parser.go

```diff
@@ -702,6 +662,112 @@ func (p *Parser) g2AppendComments(node *TreeNode) {
 	p.g2Comments = nil
 }
 
+// g2ParseBlock parses a block and its children into the given node.
+// The blockType of the node will be set to the type of the block.
+func (p *Parser) g2ParseBlock(node *TreeNode) error {
+	tok, err := p.next()
+	if err != nil {
+		return err
+	}
+
+	// Set BlockType
+	switch tok.(type) {
+	case *token.BlockStart:
+		node.Block(BlockNormal)
+	case *token.GroupStart:
+		node.Block(BlockGroup)
+	case *token.GenericStart:
+		node.Block(BlockGeneric)
+	default:
+		return token.NewPosError(tok.Pos(), "expected a BlockStart")
+	}
+
+	// Parse children
+	for {
+		if err := p.g2EatComments(); err != nil {
+			return err
+		}
+
+		p.g2AppendComments(node)
+
+		tok, err = p.peek()
+		if err != nil {
+			return err
+		}
+
+		if node.isClosedBy(tok) {
+			p.next() // pop closing token
+
+			break
+		} else if tok.TokenType() == token.TokenG2Arrow {
+			if err := p.g2ParseArrow(node); err != nil {
+				return err
+			}
+		} else if tok.TokenType() == token.TokenDefineElement {
+			children, err := p.g1LineNodes()
+			if err != nil {
+				return err
+			}
+			node.AddChildren(children...)
+		} else {
+			child, err := p.g2Node()
+			if err != nil {
+				return err
+			}
+
+			node.AddChildren(child)
+		}
+	}
+
+	return nil
+}
+
+// g2ParseArrow is used to parse the return arrow, which has special semantics.
+// It is used to append a "ret" element containing function return values to a
+// function definition. For this to work, the function must be defined as:
+//     func name(...) -> (...)
+// The "func" element can be named differently, but it must be there, as the
+// "ret" element will be placed inside it. "ret" will contain everything defined
+// in the block after "->".
+// The blocks "(...)" are required, but can be any valid blocks.
+func (p *Parser) g2ParseArrow(node *TreeNode) error {
```

Comment:
Warum ist das eigentlich ein spezielles Element und kann nicht einfach als neues Kind-Element gehandelt werden? Ich hatte eigentlich nur Klammern in g2 mit spezieller Bedeutung vorgesehen. Hab ich was übersehen, was es besonders macht?

Reply:
Ein Block ist immer das Ende einer verschachtelten Kette, Beispiel:
```
// Tadl
a b(...) c
// XML
<a>
    <b>...</b>
</a>
<c />
```
Ohne die Definition lässt sich der Pfeil einfach als "ret" Element behandeln, mein Beispiel würde sich aber verändern:
```xml
<a>
    <b>...</b>
    <c />
</a>
```
Was meiner Meinung nach nicht mehr ganz so intuitiv ist. Soll ich das ändern?
Unabhängig davon kann man den Pfeil schon noch allgemeiner betrachten, aber ich dachte, dass die Form `func name(...) -> (...)` dafür sorgt, dass der Pfeil nicht übermäßig verwendet wird. Wie würde man z.B. `a -> b` oder `a -> b -> c` parsen?

Reply:
hm, ich dachte halt einfach verschachtelt, also
```
//tadl
func name(...) -> (...)

//xml

<func>
  <name>
      ...
     <ret _type="->">
        ...
     </ret>
  </name>
</func>
```
Hätte zumindest den Vorteil, dass man alle Symbole immer gleich benutzen kann, also = oder := oder ≠ oder => oder --> oder "return" etc

Oder wir führen neben den Klammern tatsächlich weitere Operatorprioritäten ein. Dann sollten wir aber auch überlegen, ob wir die Liste nicht sogar um eine ganze Reihe von Operatoren erweitern (z.B. übliche binäre Operatoren +/-* sowie unäre etc.)

Reply:
Klingt sinnvoll. Sollte dann das Element vor dem Pfeil einen Block voraussetzen? Wäre eigentlich nicht notwendig. `name(...) -> (...)` würde mit einem leeren Block dann genauso wie `name -> (...)` geparst werden, also `<name><ret>...</ret></name>`.

Reply:
ohne Block scheint mir das auch ok, ist ja quasi Map-Style